Add Status Update packet code (#105)
vacp2p/rfc: waku/waku.md

required-packet = 0 status /
                  1 messages /

> 		  2 status-update /

Reviewer: Still need to add the `status-update` higher up. Basically containing just `status-options`.

> 		  2 status-update /

Reviewer 2: We want something similar (or the same?) as this in ABNF:

```
status-options = "["
  [ pow-requirement-key pow-requirement ]
  [ bloom-filter-key bloom-filter ]
  [ light-node-key light-node ]
  [ confirmations-enabled-key confirmations-enabled ]
  [ rate-limits-key rate-limits ]
  [ topic-interest-key topic-interest ]
"]"
````

Reviewer 2: E.g.

`status-update = status-options` (assuming that is accurate)

Author: updated

| -------------------------- | ------------- |
| Status                     |     0         |
| Messages                   |     1         |

> | Status Update              |     2         |

Reviewer 2: Could you add this and remove the other ones from ABNF form too?

This packet is used for sending the standard Waku envelopes.

#### Status Update

The Status Update message is used to communicate an update of the settings of the node.
The format is the same as the Status message, all fields are optional.
Fields that are omitted are considered unchanged.

> 

Reviewer: Should we still add here something that it is advised not to add fields that don't require an update?

> ##### PoW Requirement update

Reviewer 2: Too many levels of headers, can we replace the `#####` with `**Bold subsection:***`? Easier to skim

Reviewer: Or move that to its own section explaining the different fields of `status-options`.

> When PoW is updated, peers should no longer deliver the sender messages with PoW lower than specified in this message.

Reviewer: ```suggestion
When PoW is updated, peers SHOULD no longer deliver the sender envelopes with PoW lower than specified in this message.
```

	D[n] = 1
	END FOR

> A full bloom filter means that the node is to be considered a `Full Node` and it will accept any topic.

Reviewer 2: Do we want to specify what a full bloom filter looks like?

Reviewer 2: Also consistency: Bloom Filter vs Bloom filter etc - pick one, if field use `field` or similar. `Full node` should probably just be Full Node?

If both Topic Interest and Bloom filter are specified, Topic Interest always takes precedence and Bloom Filter MUST be ignored.

> If only Bloom Filter is specified, the current Topic Interest MUST be discarded and only the updated Bloom Filter MUST be used when forwarding envelopes.

Reviewer: ```suggestion
If only Bloom Filter is specified, the current Topic Interest MUST be discarded and only the updated Bloom Filter MUST be used when forwarding or posting envelopes.
```

> An bloom filter of all 0s signals that the node is not currently interested in receiving any envelope.

Reviewer: ```suggestion
A bloom filter with all bits set to 0 signals that the node is not currently interested in receiving any envelope.
```

Reviewer 2: Might make sense sense to state that all set to 1s indicate full node here

> ##### Topic Interest update (experimental)

Reviewer 2: I suppose this is no longer experimental

> This packet is used by Waku nodes for sharing their interest in messages with specific topics. It does this in a more bandwidth considerate way, at the expense of metadata protection. Peers MUST only send envelopes with specified topics.

Reviewer 2: ```suggestion
This packet is used by Waku nodes for sharing their interest in messages with specific topics. It does this in a more bandwidth considerate way, at the expense of some metadata protection. Peers MUST only send envelopes with specified topics.
```

> This feature will likely stop being experimental in v1.

Reviewer 2: ```suggestion
```

It is currently bounded to a maximum of 10000 topics. If you are interested in more topics than that, this is currently underspecified and likely requires updating it. The constant is subject to change.

> If both Topic Interest and Bloom filter are specified, Topic Interest always takes precedence and Bloom Filter MUST be ignored.

Reviewer: Not sure if we need to say this twice in the spec

> If only Topic Interest is specified, the current Bloom Filter MUST be discarded and only the updated Topic Interest MUST be used when forwarding envelopes.

Reviewer: ```suggestion
If only Topic Interest is specified, the current Bloom Filter MUST be discarded and only the updated Topic Interest MUST be used when forwarding or posting envelopes.
```

An empty array signals that the node is not currently interested in receiving any envelope.

> #### Rate Limits update

Reviewer 2: Heading seems off, see bold case comment above

The drawback of sending message confirmations is that it increases the noise in the network because for each sent message, a corresponding confirmation is broadcasted by one or more peers.

#### P2P Request

> This packet is used for sending Dapp-level peer-to-peer requests, e.g. Waku Mail Client requesting old messages from the Waku Mail Server (./mailserver.md).

Reviewer 2: ```suggestion
This packet is used for sending Dapp-level peer-to-peer requests, e.g. Waku Mail Client requesting old messages from the Waku Mail Server [mailserver](./mailserver.md).
```

> This packet is used for sending Dapp-level peer-to-peer requests, e.g. Waku Mail Client requesting old messages from the Waku Mail Server (./mailserver.md).

Reviewer 2: ```suggestion
This packet is used for sending Dapp-level peer-to-peer requests, e.g. Waku Mail Client requesting old messages from the [Waku Mail Server](./mailserver.md).
```

The drawback of sending message confirmations is that it increases the noise in the network because for each sent message, a corresponding confirmation is broadcasted by one or more peers.

#### P2P Request

> This packet is used for sending Dapp-level peer-to-peer requests, e.g. Waku Mail Client requesting old messages from the Waku Mail Server.

Reviewer 2: What do we mean by Dapp-level here?

Also please link to mailserver spec

Author: Not sure, this section has only been moved @adambabik

## Changelog

### Version 0.4

Released [February 20, 2020](https://github.com/vacp2p/specs/commit/dafe08ea8792fbb9afa0be67cda3c97c79a1de67).

> - Add Status Update code `0x02`

Reviewer 2: Is this 0x02 or 0x22? Maybe I fat fingered it with suggestion, right now there is one too many lines anyway

Author: `0x2` I'll edit it

- Introduces a new packet code Status Code (`0x02`) for communicating option changes

> - Breaking: Removes support for packet codes `0x02`, `0x03`, `0x20`; superseded by `0x02`

Reviewer 2: ```suggestion
- Breaking: Removes support for the following packet codes: PoW Requirement (`0x02`), Bloom Filter (`0x03`), Rate limits (`0x20`), Topic interest (`0x21`) - all superseded by the new Status Code (`0x02`)
```

Reviewer 2: Re-use of packet codes made it read a bit weird so made it more explicit for clarity

Reviewer 2: ^ @cammellos

## Changelog

> ### Version 0.4

Reviewer 2: In follow up commit let's add Release <date> with link as a signed commit and tag

> - Add Status Update code 0x02

Reviewer 2: ```suggestion
- Add Status Update code `0x02`
```

And similar below

> - Removes old 0x02, 0x03, 0x20 as superseded

Reviewer 2: ```suggestion
- Introduces a new packet code Status Code (`0x22`) for communicating option changes
- Breaking: Removes support for packet codes `0x02`, `0x03`, `0x20`; superseded by `0x22`
```

> - Increased topic interest to 10000

Reviewer 2: ```suggestion
- Increased `topic-interest` capacity from 1000 to 10000
```
